### ISSUE TITLE
Update edgeOtsu method to allow for empty histogram

### ISSUE THREAD

**Issue by ArjenHaag** [Servir-Mekong/hydra-floods#11]
Otsu's method requires a histogram that contains data. When very little or no water is present in the image that is used to construct the histogram, it will fail. This can be overcome by setting a (quite low, e.g. for Sentinel-1 something like -18) static threshold when this occurs.

From JavaScript version:
`var threshold = ee.Number(ee.Algorithms.If(histogram.contains('bucketMeans'), otsu(histogram), thresh_no_data))`

ee.Algorithms.If is perhaps not the best way to do this, so while this will definitely work, it would also be good to think whether it can be done in a better way.

**Comment by ArjenHaag** [comment]
@KMarkert Just added a quick fix, still need to test it. However, also have a few questions:

1. should this fix also be added to the bmax method?
2. I've now set the default threshold for when there is no histogram at -0.2, since the other defaults seem to be tuned to optical indices (e.g. initial threshold at 0). Is that OK?
3. I've not seen it being used anywhere except at the fused workflow, what would be a good threshold there?

And finally, more of a general question, it might be good to store default parameters for each satellite/sensor (type) somewhere? What do you think?

**Comment by KMarkert** [comment]
Thanks for tackling this! Given that `ee.Algortihms.If` is not so efficient, maybe we can set the default no data value to None only do the check if the users explicitly calls it? Like this:
```python
if thresh_no_data is not None:
    threshold = ee.Number(ee.Algorithms.If(histogram.contains('bucketMeans'), otsu(histogram), thresh_no_data))
```
Just a thought, however, this may not be super intuitive for users...

For the other questions here are my answers:

1.  Yes, i think it would be great if we could have this added to the bmax method as well
2. Yes, this works (unless we want to set it as None like discussed above). Not sure why I had those initial values but keeping them consistent seems best.
3. For the fusion process I think setting the `thresh_no_data` to -0.1 or -0.15 would be best. The range is a little different than a regular optical water index. We would probably need to check to make sure though...

I like the idea of creating lookups for default parameters! When setting default function parameters per sensor there could be two options.

Option 1: create a `defaults.py` module and add keywords as dictionaries. This way users can quickly import that module and get the suggested function parameters per sensor. Something like:
```python
s1 = {
    'edge_otsu': {
        'initial_threshold':-16,
        'edge_buffer': 300,
         ...
    },
    ...
}

'landsat' = {'edge_otsu':{...},...}
```
Pro would be that users can access the defaults easily and pass directly to the functions whether using the datasets classes, image collections, or simply images (i.e. `water = hf.edge_otsu(s1_img, **defaults.s1['edge_otsu'])`). Con would it can be that the sensor lookup variable names may be different than what a user is expecting/knows.

Option 2: include these defaults within the sensor dataset classes as object parameters, see [`datasets.py`](https://github.com/Servir-Mekong/hydra-floods/blob/empty-hist/hydrafloods/datasets.py). Pro would be if a user is working directly with the dataset objects they can call the defaults directly from there (i.e. `water = s1.apply_func(hf.edge_otsu, **s1.edge_otsu_defaults)`). Con would be users would have to use the dataset classes to get the defaults.

Let me know your thoughts. Again I feel like this would be very helpful, especially to new users that may not know what parameters to use.

**Comment by ArjenHaag** [comment]
Great, thanks for the detailed answers! I'll make a new issue for the default parameters as we don't clutter this one. I'll update the code based on your answers and think a bit more on the best way to use/set that threshold for when there is no data.

**Comment by ArjenHaag** [comment]
@KMarkert, perhaps one idea to circumvent the use of `ee.Algorithms.If`, although it is a bit far-fetched:

Always create a default Dictionary with just a little bit of data, which when passed into the Otsu function will return a sufficiently low value (or even the passed thresh_no_data, but that would require backtracking to construct a histogram specifically for that value). Then put both the regular and this new default Dictionary into a Feature, assign it a property based on the number of keys (empty histogram will have no keys, so should be zero), and combine those into a FeatureCollection. This can then be filtered by extracting the Feature (and from that the Dictionary, i.e. histogram) that has the largest value for that property.

Something like:

```python
# construct default histogram
default_hist = ee.Dictionary({
    bucketMeans: [..., ...],
    histogram: [..., ...]
})
# get current histogram
current_hist = histogram.get(histBand.cat("_histogram"))
# place histograms into Features and set property for filtering
default_f = ee.Feature(default_hist).set('hist_size', default_hist.keys().size())  # size of 2
current_f = ee.Feature(current_hist).set('hist_size', current_hist.keys().size())  # size of 0 when empty, 4 when not
# place into FeatureCollection and obtain non-empty histogram
fc_hists = ee.FeatureCollection([default_f, current_f])
non_empty_hist = ee.Feature(fc_hists.filter(ee.Filter.gte('hist_size', 2)).first()).toDictionary()
# derive threshold using Otsu
threshold = otsu(non_empty_hist)
```

**Comment by KMarkert** [comment]
Closed with PR #16. Conversation on `thresh_no_data` vs `default_hist` has been moved to #17.
